Big pass on opacity, blending, and clipping (#288)
browserengineering/book: book/visual-effects.md

behave just as it did in previous chapters, and it'll probably feel
faster, because Skia and SDL are faster than Tkinter.

Let's reward ourselves for the big refactor with a simple feature that
Skia enables: rounded corners via the `border-radius` CSS property,
like this:

    <div style="border-radius: 10px">
        This is some example text.
    </div>

Which looks like this:

<div style="border-radius: 10px">
This is some example text.
</div>

> Implementing requires drawing a rounded rectangle, so let's add a new

Reviewer: I'll resolve this later.

`Save`, and you should always prefer `Save` to `SaveLayer`, all things being
equal.

Compositing with SaveLayer
==========================

Color mixing happens when multiple page elements overlap. The easiest
way that happens in our browser is child elements overlapping their
parents, like this:[^transforms-etc]

[^transforms-etc]: There are way more ways elements can overlap in a
real browser: the `transform` property, `position`ed elements,
negative margins, and so many more. But color mixing works the same
way each time.

``` {.html.example}
<div style="background-color:orange">
    Parent
    <div style="background-color:white">Child</div>
    Parent
</div>
```

It looks like this:

<div style="background-color:orange">
Parent
<div style="background-color:white">Child</div>
Parent
</div>

Right now, the white rectangle completely obscures part of the orange
one; The two colors don't really need to "mix", and in fact it kind of
looks like two orange rectangles instead of an orange rectangle with a
white one on top. Now let's make the white child element
semi-transparent, so the colors have to mix. In CSS, that requires

> adding an `opacity` property with a value somewhere in between 0

Reviewer: ```suggestion
adding an `opacity` property with a value somewhere between 0
```

> (completely transparent) to 1 (totally opaque). With 50% opacity on

Reviewer: ```suggestion
(completely transparent) and 1 (totally opaque). With 50% opacity on
```

        if bgcolor != "transparent":
            cmds.append(DrawRect(rect, bgcolor))

        for child in self.children:
            child.paint(cmds)
        
        display_list.extend(cmds)
```

Now, _before_ we add our temporary command list to the overall display
list, we can use `SaveLayer` to add transparency to the whole element.
I'm going to do this in a new `paint_visual_effects` method, because
we'll want to make the same changes to all of our other layout
objects:

``` {.python}
class BlockLayout:
    def paint(self, display_list):
        # ...
        cmds = paint_visual_effects(self.node, cmds, rect)
        display_list.extend(cmds)
```

Inside `paint_visual_effects`, we'll parse the opacity value and
construct the appropriate `SaveLayer`:

``` {.python}
def paint_visual_effects(node, cmds, rect):
    opacity = float(node.style.get("opacity", "1.0"))
    paint = skia.Paint(Alphaf=opacity)
    cmds = [SaveLayer(paint, cmds)]
    return cmds
```

Note that `paint_visual_effects` receives a list of commands and
returns another list of commands. It's just that the output list is
always a single `SaveLayer` command.

Compositing Pixels
==================

> Let's pause and explore how opacity actually works under the hood.

Reviewer: ```suggestion
Now let's pause and explore how opacity actually works under the hood.
```

*premultiplied* representation of the color.

[^alpha-history]: Check out this [history of alpha][alpha-history],
written by its co-inventor (and co-founder of Pixar), or read this
[derivation of alpha][alpha-deriv] for how it is computed.

[alpha-history]: http://alvyray.com/Memos/CG/Microsoft/7_alpha.pdf
[alpha-deriv]: https://jcgt.org/published/0004/02/03/paper.pdf

When a pixel with alpha overlaps another pixel, the final color is a
mix of their two colors. How exactly the colors are mixed is defined
by Skia's `Paint` objects. Of course, Skia is pretty complex, but we
can sketch these paint operations in Python as methods on an imaginary
`Pixel` class.

``` {.python file=examples}
class Pixel:
    def __init__(self, r, g, b, a):
        self.r = r
        self.g = g
        self.b = b
        self.a = a
```

When we apply a `Paint` with an `Alphaf` parameter, the first thing
Skia does is add the requested opacity to each pixel:

``` {.python file=examples}
class Pixel:
    def alphaf(self, opacity):
        self.a = self.a * opacity
```

I want to emphasize that this code is not a part of our browser---I'm
simply using Python code to illustrate what Skia is doing internally.

> That `Alphaf` operation applies to pixel in one surface. But with

Reviewer: ```suggestion
That `Alphaf` operation applies to pixels in one surface. But with
```

`SaveLayer` we will end up with two surfaces, with all of their pixels

> aligned, and therefore we will need to combined, or *blend* pairs of

Reviewer: ```suggestion
aligned, and therefore we will need to combine, or *blend* pairs of
```

corresponding pixels.

Here the terminology can get confusing: we imagine that the pixels "on
top" are blending into the pixels "below", so we call the top surface
the *source surface*, with source pixels, and the bottom surface the

> *destination surface*, with destination pixels. When we combine them,

Reviewer: I called it backdrop to conform to the spec. If destination is more clear, we can use that. But backdrop
does have an intuitive meaning maybe?

Author: Yeah I could go either way here. "Destination" is nice because then "destination-in" makes a lot of sense. But "backdrop" is nice because it reminds you which one is the backdrop. We could also introduce both terms, but that itself can be confusing.

Reviewer: ok let's go with destination because of dst-in.

there are lots of ways we could do it, but the default on the web is
called "simple alpha compositing" or *source-over* compositing. In
Python, the code to implement it looks like this:[^simple-alpha]

[^simple-alpha]: The formula for this code can be found
[here](https://www.w3.org/TR/SVG11/masking.html#SimpleAlphaBlending).
Note that that page refers to premultiplied alpha colors, but Skia

> does not use premultiplied color representations internally.

Reviewer: ```suggestion
does not use premultiplied color representations internally, and the code below doesn't either.
```

``` {.python file=examples}
class Pixel:
    def source_over(self, source):
        self.r = self.r * (1 - source.a) * self.a + source.r * source.a
        self.g = self.g * (1 - source.a) * self.a + source.g * source.a
        self.b = self.b * (1 - source.a) * self.a + source.b * source.a
        self.a = 1 - (1 - source.a) * (1 - self.a))
```

Here the destination pixel `self` is modified to blend in the source
pixel `source`. The mathematical expressions for the red, green, and
blue color channels are identical, and basically average the source

> and destination colors, weighted by alpha.[^source-over-example] You

Reviewer: Footnote missing?

``` {.python}
class BlockLayout:
    def paint(self, display_list):
        cmds = []

        # ...
        if bgcolor != "transparent":
            cmds.append(DrawRect(rect, bgcolor))

        for child in self.children:
            child.paint(cmds)
        
        display_list.extend(cmds)
```

Now, _before_ we add our temporary command list to the overall display
list, we can use `SaveLayer` to add transparency to the whole element.
I'm going to do this in a new `paint_visual_effects` method, because
we'll want to make the same changes to all of our other layout
objects:

``` {.python}
class BlockLayout:
    def paint(self, display_list):
        # ...
        cmds = paint_visual_effects(self.node, cmds, rect)
        display_list.extend(cmds)
```

Inside `paint_visual_effects`, we'll parse the opacity value and
construct the appropriate `SaveLayer`:

``` {.python}
def paint_visual_effects(node, cmds, rect):
    opacity = float(node.style.get("opacity", "1.0"))

    return [
         SaveLayer(skia.Paint(Alphaf=opacity), cmds),
    ]
```

Note that `paint_visual_effects` receives a list of commands and
returns another list of commands. It's just that the output list is

> always a single `SaveLayer` command.

Reviewer: ```suggestion
always a single `SaveLayer` command that wraps the original content---which makes sense, because
first we need to draw the commands to a surface, and *then* apply transparency to it when blending into
the parent.
```

for this use case).
:::

::: {.todo}
<!-- There's nothing for this comment to describe anymore, but it's -->
<!-- useful information so I'm keeping it in a todo block for now-->
Here I had to explicitly set a background color of white on the
`<html>` element, even though web pages have a default white background. This
is because `mix-blend-mode` is defined in terms of stacking contexts. In a real
browser, if a stacking context doesn't paint anything, then its blending
surface is empty; in our browser, blending happens whith whatever surface
happened to be there before `saveLayer` was called. This is a bug in our
browser, which can be fixed by calling `saveLayer` on the parent layout
object.
:::

Clipping and masking
====================

The "multiply" and "difference" blend modes can seem kind of obscure,
but blend modes are a flexible way to implement per-pixel operations.
One common use case is clipping---erasing the pixels in a surface that
don't intersect with a given shape. It's called clipping because it's
like putting a second piece of paper (called a *mask*) over the first
one, and then using scissors to cut along the mask's edge.

> Clipping can be used via the CSS `clip-path` property. The [full

Reviewer: ```suggestion
One way of expressing clipping on the web is via the CSS `clip-path` property. The [full
```

definition](https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path)
is quite complicated, so let's just implement circular clips via the
`circle(XXpx)`, which is used like this:

``` {.html.example}
<div style="clip-path:circle(20px);background-color:lightblue">
    This test text exists here to ensure that the "div" element is
    large enough that a reasonably large circle is drawn to your screen.
</div>
```

That HTML looks like this:

<div style="clip-path:circle(20px);background-color:lightblue">
This test text exists here to ensure that the "div" element is
large enough that a reasonably large circle is drawn to your screen.
</div>

> Here, the percentage defines the radius of the circle, and is

Reviewer: Remove this paragraph.

        return skia.BlendMode.kDifference
    else:
        return skia.BlendMode.kSrcOver
```

This makes adding support for blend modes to our browser as simple as
passing the `BlendMode` parameter to the `Paint` object:

``` {.python}
def paint_visual_effects(node, cmds, rect):
    # ...
    blend_mode = parse_blend_mode(node.style.get("mix-blend-mode"))
    
    return [
        SaveLayer(skia.Paint(BlendMode=blend_mode), [
            SaveLayer(skia.Paint(Alphaf=opacity), cmds),
        ]),
    ]
```

Note the order of operations here: we _first_ apply transparency, and
_then_ blend the result into the rest of the page. If we switched the
two `SaveLayer` calls, so that we first applied blending, there
wouldn't be anything to blend it into!

[mbm-mult]: https://drafts.fxtf.org/compositing-1/#blendingmultiply
[mbm-diff]: https://drafts.fxtf.org/compositing-1/#blendingdifference

::: {.further}

> In reality, reading individual pixels into memory and manipulate them

Reviewer: ```suggestion
In reality, reading individual pixels into memory to manipulate them
```

surface is the content we want to clip, so destination-in fits
perfectly. In code, destination-in looks like this:

[dst-in]: https://drafts.fxtf.org/compositing-1/#porterduffcompositingoperators_dstin

``` {.python file=examples}
class Pixel:
    def destination_in(self, source):
        self.r = self.r * self.a * source.a
        self.g = self.g * self.a * source.a
        self.b = self.b * self.a * source.a
        self.a = self.a * source.a
```

To implement this, we'll need to be able to draw a rounded rectangle:

``` {.python}
class DrawRRect:
    def __init__(self, rect, radius, color):
        self.rect = rect
        self.rrect = skia.RRect.MakeRectXY(rect, self.radius, self.radius)
        self.color = color

    def execute(self, canvas):
        canvas.drawRRect(self.rrect, paint=skia.Paint(Color=self.color))
```

Now, in `paint_visual_effects`, we need to create a new layer, draw
the mask image into it, and then blend it with the element contents
with destination-in blending:

``` {.python}
def paint_visual_effects(node, cmds, rect):
    # ...
    border_radius = float(node.style.get("border-radius", "0px")[:-2])

>     return [

Reviewer: \o/

let's focus here on `overflow: clip`, which cuts off contents of an
element that are outside the element's bounds.

[^like-clip-path]: The CSS [`clip-path` property][mdn-clip-path] lets
specify a mask shape using a curve, while the [`mask`
property][mdn-mask] lets you instead specify a image URL for the mask.

[^like-scroll]: For example, `overflow: scroll` adds scroll bars and
    makes an element scrollable, while `overflow: hidden` is similar
    but subtly different from `overflow: clip`.

[mdn-mask]: https://developer.mozilla.org/en-US/docs/Web/CSS/mask

[mdn-clip-path]: https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path

Usually, `overflow: clip` is used with properties like `height` or
`rotate` which can make an element's children poke outside their
parent. Our browser doesn't support these, but there is one edge case
where `overflow: clip` is relevant: with rounded corners. Consider
this example:

``` {.html.example}
<div style="border-radius:30px;background-color:lightblue;overflow:clip">
    This test text exists here to ensure that the "div" element is
    large enough that the border radius is obvious.
</div>
```

That HTML looks like this:

> <div style="border-radius:30px;background-color:lightblue;overflow:clip">

Reviewer: We can add overflow:hidden as a hack. Also, I happen to know that the implementation is complete and it'll ship early next year...

Reviewer: I'll resolve this later.

[overflow-prop]: https://developer.mozilla.org/en-US/docs/Web/CSS/overflow

Optimizing Surface Use
======================

While the `mask` CSS property is relatively uncommon used (as is `clip-path`
actually), there is a special kind of mask that is very common: rounded
corners. Now that we know how to implement masks, this one is also easy to
add to our browser (draw a mask image with the `drawRRect` Skia canvas method).
But because it's so common, Skia provides a special-purpose
method to clip to rounded corners: `clipRRect`.

Rounded corners are specified in CSS via `border-radius`. Here's an example:

>     <div style="width:256px; height:256px;

Reviewer: Change to not use width or height, but instead text, to make it work in our browser?

Author: Yes. I haven't passed over this section yet but definitely switch to text.
